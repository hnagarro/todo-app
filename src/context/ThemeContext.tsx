import React, { createContext, useState, PropsWithChildren } from 'react';

type Theme = 'light' | 'dark';

interface ThemeContextType {
  theme: Theme;
  toggleTheme: () => void;
}

export const ThemeContext = createContext<ThemeContextType>({
  theme: 'light',
  toggleTheme: () => {},
});

export const ThemeProvider: React.FC<PropsWithChildren<{}>> = ({ children }) => {
  const [theme, setTheme] = useState<Theme>('light');

  const toggleTheme = () => {
    setTheme(theme === 'light' ? 'dark' : 'light');
  };

  return (
    <ThemeContext.Provider value={{ theme, toggleTheme }}>
      {children}
    </ThemeContext.Provider>
  );
};
