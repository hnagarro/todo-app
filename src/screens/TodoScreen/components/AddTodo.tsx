import React from 'react';
import { View, TextInput, StyleSheet } from 'react-native';
import { useTranslation } from 'react-i18next';
import { ThemeContext } from '../../../context/ThemeContext';
import Button from '../../../components/Button';
import { useSelector } from 'react-redux';
import { RootState } from '../../../redux/rootReducer';

interface AddTodoProps {
  text: string;
  onChangeText: (text: string) => void;
  onAddTodo: () => void;
}

const AddTodo: React.FC<AddTodoProps> = ({ text, onChangeText, onAddTodo }) => {
  const { t } = useTranslation();
  const { theme } = React.useContext(ThemeContext);
  const loading = useSelector((state: RootState)=> state.todos.loading)
  return (
    <View style={styles.addTodoContainer} >
      <TextInput
        style={[styles.input, theme === 'dark' && styles.darkInput]}
        placeholder={t('enter_todo')}
        value={text}
        onChangeText={onChangeText}
        placeholderTextColor={theme === 'dark' ? "#fff" : "#000"} 
        testID='enter_todo'
      />
      <Button title={t('add_todo')} testID='add_todo' onPress={onAddTodo} disabled={loading}/>
    </View>
  );
};

const styles = StyleSheet.create({
  addTodoContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    marginBottom: 20
  },
  input: {
    borderWidth: 1,
    borderColor: '#ccc',
    padding: 10,
    flex: 1,
    color: '#000',
  },
  darkInput: {
    backgroundColor: '#444',
    color: '#fff',
    borderColor: '#555',
  },
});

export default AddTodo;
