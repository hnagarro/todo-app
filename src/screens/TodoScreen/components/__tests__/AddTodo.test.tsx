import React from 'react';
import { render } from '@testing-library/react-native';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { ThemeContext } from '../../../../context/ThemeContext';
import AddTodo from '../AddTodo';

jest.mock('react-i18next', () => ({
  useTranslation: () => ({
    t: (key: any) => key,
    i18n: {
      changeLanguage: () => new Promise(() => {}),
    },
  }),
}));

const mockStore = configureStore([]);
const initialState = {
  todos: { loading: false },
};

describe('AddTodo Component', () => {
  let store:any;
  const mockOnChangeText = jest.fn();
  const mockOnAddTodo = jest.fn();

  beforeEach(() => {
    store = mockStore(initialState);
  });

  it('should render correctly with dark theme', () => {
    const { getByPlaceholderText } = render(
      <Provider store={store}>
        <ThemeContext.Provider value={{ theme: 'dark', toggleTheme: ()=>{} }}>
          <AddTodo text="" onChangeText={mockOnChangeText} onAddTodo={mockOnAddTodo} />
        </ThemeContext.Provider>
      </Provider>
    );

    const input = getByPlaceholderText('enter_todo');

    expect(input.props.placeholderTextColor).toBe('#fff');
    expect(input.props.style).toContainEqual({ backgroundColor: '#444', color: '#fff', borderColor: '#555' });
  });

  it('should disable button when loading is true', () => {
    store = mockStore({
      todos: { loading: true },
    });

    const { getByText, toJSON, getByTestId } = render(
      <Provider store={store}>
        <ThemeContext.Provider value={{ theme: 'light', toggleTheme: ()=>{} }}>
          <AddTodo text="new todo" onChangeText={mockOnChangeText} onAddTodo={mockOnAddTodo} />
        </ThemeContext.Provider>
      </Provider>
    );

    const button = getByTestId('add_todo');
    expect(toJSON()).toMatchSnapshot('button snapshot')
  });
});
