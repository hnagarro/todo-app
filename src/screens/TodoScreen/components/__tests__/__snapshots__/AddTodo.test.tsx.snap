// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTodo Component should disable button when loading is true: button snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginBottom": 20,
    }
  }
>
  <TextInput
    onChangeText={[MockFunction]}
    placeholder="enter_todo"
    placeholderTextColor="#000"
    style={
      [
        {
          "borderColor": "#ccc",
          "borderWidth": 1,
          "color": "#000",
          "flex": 1,
          "padding": 10,
        },
        false,
      ]
    }
    testID="enter_todo"
    value="new todo"
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": true,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#d3d3d3",
        "borderRadius": 5,
        "justifyContent": "center",
        "marginLeft": 10,
        "opacity": 1,
        "paddingHorizontal": 20,
        "paddingVertical": 10,
      }
    }
    testID="add_todo"
  >
    <Text
      style={
        [
          {
            "color": "#fff",
            "fontSize": 16,
          },
          undefined,
        ]
      }
    >
      add_todo
    </Text>
  </View>
</View>
`;
